[MRG + 1] FIX raise an error message when n_groups > number of groups (#7681)

polmauri opened the pull request:
This change fixes issue #7681:
- Raise `ValueError` when `n_groups` > actual number of unique groups in `LeaveOneGroupOut` and `LeavePGroupsOut`.
- Add unit test.

polmauri committed:
FIX raise an error message when n_groups > actual number of groups (#7681)

This change addresses issue #7681:
- Raise ValueError when n_groups > actual number of unique groups in LeaveOneGroupOut and LeavePGroupsOut.
- Add unit test.

nelson-liu commented on sklearn/model_selection/_split.py:
```diff
@@ -862,6 +864,10 @@ def _iter_test_masks(self, X, y, groups):
         unique_groups = np.unique(groups)
+        if self.n_groups > len(unique_groups):
+            raise ValueError("Cannot have n_groups=%d greater than the number "
+                             "of unique groups: %d."
```
i find it odd that you choose to use `=` to indicate the variable value in for n_groups and `:`  for the number of unique groups; maybe keep it consistent?

polmauri replied:
I copied the format from this other [error](https://github.com/polmauri/scikit-learn/blob/5a37b37958438207393d93fa6774cc441a97ccb2/sklearn/model_selection/_split.py#L868) in the same file, should I change it?

tguillemot replied:
If all the file use `=`, you can keep `=`.

raghavrv replied:
I think it's okay as `n_groups` is a parameter and unique groups is the information we are trying to include in the error message...

raghavrv replied:
--> `>=`

raghavrv commented:
Thanks for the PR @polmauri

raghavrv commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -724,6 +724,19 @@ def test_leave_group_out_changing_groups():
+        list(LeavePGroupsOut(n_groups=5).split(X, y, groups))
+
+    assert_raises(ValueError, leave_one_group_out_zero_groups)
+    assert_raises(ValueError, leave_p_groups_out_greater_than_unique_groups)
```
Could you instead test these like

```
assert_raise_message(ValueError, "<snippet of the error message">,
                     LeaveOneGroupOut().split, X, y)
```

raghavrv commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -724,6 +724,19 @@ def test_leave_group_out_changing_groups():
     assert_equal(3, LeaveOneGroupOut().get_n_splits(X, y, groups))
 
 
+def test_leave_group_out_split_errors():
```
--> `test_leave_one_p_group_out_error_on_fewer_number_of_groups`

tguillemot commented:
Thanks @polmauri. 
+1 once corrected.

polmauri commented:
Changed:
- Check error message with `assert_raise_message`
- Pass parameters to `assert_raise_message` instead of defining functions

polmauri committed:
Make requested changes

- Check error message with `assert_raise_message`
- Pass parameters to `assert_raise_message` instead of defining functions

raghavrv commented on sklearn/model_selection/_split.py:
```diff
@@ -773,6 +773,8 @@ def _iter_test_masks(self, X, y, groups):
         # We make a copy of groups to avoid side-effects during iteration
         groups = np.array(groups, copy=True)
         unique_groups = np.unique(groups)
+        if len(unique_groups) == 0:
```
`len(unique_groups) <= 1`

amueller replied:
needs a test.

raghavrv commented on sklearn/model_selection/_split.py:
```diff
@@ -773,6 +773,8 @@ def _iter_test_masks(self, X, y, groups):
         groups = np.array(groups, copy=True)
         unique_groups = np.unique(groups)
+        if len(unique_groups) == 0:
+            raise ValueError("Cannot have zero groups")
```
```python
"The groups parameter contains fewer than 2 unique groups (%s). `LeaveOneGroupOut` expects at least 2."
% (unique_groups)
```

raghavrv commented on sklearn/model_selection/_split.py:
```diff
@@ -862,6 +864,10 @@ def _iter_test_masks(self, X, y, groups):
             raise ValueError("The groups parameter should not be None")
         groups = np.array(groups, copy=True)
         unique_groups = np.unique(groups)
+        if self.n_groups > len(unique_groups):
```
`>=`

raghavrv commented on sklearn/model_selection/_split.py:
```diff
@@ -862,6 +864,10 @@ def _iter_test_masks(self, X, y, groups):
         groups = np.array(groups, copy=True)
         unique_groups = np.unique(groups)
+        if self.n_groups > len(unique_groups):
+            raise ValueError("Cannot have n_groups=%d greater than the number "
```
```python
"The groups parameter contains fewer than (or equal to) n_groups (%d) numbers of unique groups"
" (%s). LeavePGroupsOut expects that atleast n_groups + 1 (%d) unique groups be present"
% (n_groups, unique_groups, n_groups + 1)
```

amueller commented:
@polmauri are you still working on this?

polmauri committed:
Update condition and exception message

polmauri commented:
Changes requested done.

raghavrv commented:
Thanks for addressing all the comments. Looks good to me.

amueller commented:
thanks :)
